doc/api/cli.md

```diff
@@ -11,16 +11,42 @@ To view this documentation as a manual page in a terminal, run `man node`.
 
 ## Synopsis
 
-`node [options] [V8 options] [script.js | -e "script" | -] [--] [arguments]`
+`node [options] [V8 options] [<program-entry-point> | -e "script" | -] [--] [arguments]`
 
-`node inspect [script.js | -e "script" | <host>:<port>] …`
+`node inspect [<program-entry-point> | -e "script" | <host>:<port>] …`
 
 `node --v8-options`
 
 Execute without arguments to start the [REPL][].
 
 For more info about `node inspect`, see the [debugger][] documentation.
 
+## Program entry point
+
+The program entry point is a specifier-like string. If the string is not an
+absolute path, it's resolved as a relative path from the current working
+directory. That path is then resolved by [CommonJS][] module loader. If no
+corresponding file is found, an error is thrown.
+
+If a file is found, its path will be passed to the [ECMAScript module loader][]
+under any of the following conditions:
+
+* The file has an `.mjs` extension.
+* The file nearest parent `package.json` file
+  contains a top-level [`"type"`][] field with a value of `"module"`.
+* If the program was started with a command-line flag that forces the entry
+  point to be loaded with ECMAScript module loader.
```

Comment:
I think you should explicitly refer to `--input-type`. I know that `--loader` also currently makes the entry point esm, but that's a bug (cf https://github.com/nodejs/node/issues/33226, cc @cspotcode).

Formalizing this behaviour on `--loader` is problematic because it means whoever sets must also be sure that the entry point is esm. This isn't always the case: `NODE_OPTIONS` is typically set long before `node` is ever called, and by processes which aren't necessarily aware of what the entry point will be (or even if there will only be one of them).

Reply:
To save readers some time, this is the comment where we established that it is, indeed, a bug: https://github.com/nodejs/node/issues/33226#issuecomment-791610193

Reply:
This PR is trying to describe the current behavior of Node.js, not its final or ideal state. I don't call out which CLI flags do trigger this behavior precisely because of that discussion you linked to. `--experimental-specifier-resolution=node` also triggers this behavior.

Reply:
The nature of the fix isn't clear yet so idk if this PR needs to fix it right now. I wouldn't want to block this PR based upon #33226 .

Reply:
I only meant that documenting the interaction between `--loader` and entry points (even if `--loader` isn't explicitly referred to in this paragraph) may have a slight risk of suggesting to readers that it can be relied upon and that it'd be a regression to fix it. While loaders are experimental, I noticed people often expect experimental-induced breaking changes to be very visible in their API (like the hook refactoring) rather than subtle under-the-hood behavioural changes.

Note that this wasn't a blocker by any means, just a suggestion.